scenarios.md

```diff
@@ -250,6 +250,40 @@ A weakness is discovered in a widely used cryptographic algorithm and a decision
 1. Key revocation, chain of trust, etc. must all work for the expected lifetime of a version of the client software while these changes are made.
 1. The actions that different parties need to perform must be clearly articulated, along with the result of not performing those actions.
 
+### Scenario #10: Specifying a trusted key
+
+A user may want to trust an artifact only if it is signed by a specific key. Maybe this is an internally created artifact with a known signing key. This key may be distributed using a trusted third party mechanism.
+
+1. The user obtains a trusted key for a particular artifact.
+1. The user downloads and verifies the artifact using Notary v2 and their known key.
+
+**Implications of this requirement**
+
+1. Users must be allowed to configure specific trusted keys for specific artifacts.
+
+### Scenario #11: Using a default trusted key
+
+If a user does not have a specific key for a given artifact, verified using a third party system, they will need to determine the trusted signing key(s) for an artifact using a secure default method.
+
+1. The user determines the default trusted key(s) for a specific artifact using information available on the registry, using delegations from a trusted root key.
+1. The user downloads and verifies an artifact using Notary v2 and the default trusted key(s)
+
+**Implications of this requirement**
+
+1. Users must be able to obtain per-package trusted keys, verified by a trusted root.
+
+### Scenario #12: Using multiple roots of trust
+
+A user using multiple registries will want to ensure artifacts are verified using the correct root of trust. For example, if an artifact is supposed to be signed by a key delegated to their private registry, they never want to download a version signed by a key from a public registry. Additionally, they may always want to look for artifacts signed by their private registry key first, and so want an enforced ordering of roots of trust.
```

Comment:
Is this attempting to prevent dependency confusion attacks? Since all image pulls are uniquely named, we don't have the dependency confusion attacks where a user attempts to pull an image from a private registry and accidentally downloads from a public one. For a mirrored registry, the pull is either explicit to use that mirror, or the mirror is treated as the same as the upstream registry and the risk is actually reversed (that the mirror has become stale).

As a user, I'd like to invert the logic on this. Rather than searching for the correct key for a specific repo, I'd like to be able to scope what repos I trust a key to sign. If it's an internal company key, I may trust it for all scopes. For the Docker Inc key, I may trust it for `docker.io/library/*`. And for a vendor key, I may trust for that vendor's namespace.

If a user pulls from a mirror, and checks the signature on that mirror reference, they either need to add more scopes onto the keys, or they need a way to specify registry mirrors as a shorthand to multiply the scopes. But I think that gets into implementation details that we don't need to sort out in the scenarios doc just yet.

Given this, I don't think there's a need to prioritize roots of trust. If a key is trusted for a given scope, the signature should be considered valid.

Reply:
I updated to use the inverted logic you propose, but I think that this still requires an ordering of the roots of trust that delegate these keys. If you have a root of trust for internal company keys, and a root of trust for public registry keys, you'll want to prioritize the internal root of trust, then ensure that each individual key has a well-defined scope within the scope of the root of trust.  Basically, you start by saying 'I trust this root of trust to tell me keys for these registries', then each delegation says 'use this key for this subset of registries/repositories', until you know the exact key to use for verification.

Reply:
I think the keys are decoupled from repositories, a publisher may use the same key and push artifacts across different repository/registry. Also, it's unclear why and ordering of trusted roots is required. What role does the ordering play in signature verification?

Reply:
To me, configuration ordering is a client-side configuration concept, and the (interoperability-focused) signature specification shouldn’t need to care.

Whether the client configures (with priorities and fallback and termination clauses)
```
private-registry.example:
    prio: 100
    keys: my-private-key
public-registry.com: # (when mirrored) accept my-private-key for private overrides of publicly-hosted content
    prio: 99
    keys: my-private-key
public-registry.com/vendor:
    prio: 10
    keys: vendor-private-key
    stop-search # don't try the default CA, we know the right key for this vendor
public-registry.com:
    prio: 0
    key-ca: public-registry-ca-key # let the registry provide per-repo keys
```

or (with only using the most-matching clause and ignoring everything else, potentially asking the users to copy&paste the wide-scoped keys into narrowly-scoped stanzas, OTOH being very explicit about the trusted keys):
```
private-registry.example:
    keys: my-private-key
public-registry.com/vendor:
    keys: vendor-private-key, my-private-key # (when mirrored) accept my-private-key for private overrides of publicly-hosted content
    # key-ca not specified: don't try the default CA, we know the right key for this vendor
public-registry.com:
    keys: my-private-key # (when mirrored) accept my-private-key for private overrides of publicly-hosted content
    key-ca: public-registry-ca-key # let the registry provide per-repo keys
```

is just a UX design concern, not a feature requirement: both express the same set of trusted keys.

Reply:
Rereading
> Additionally, they may always want to look for artifacts signed by their private registry key first

Is this just saying “some configuration should be processed first” as in other examples in this PR, or actually codifying the (WIP/proposed?) Notary V1 / TUF feature of signing absence, so that given prioritized (key-1, key-2), key-2 signatures for $repo are only accepted if we have a key-1 signed assertion that key-1 is not signing any content for $repo ?